Embed `expect-webdriverio` as built in assertion API

christian-bromann opened the pull request:
## Proposed changes

So far users have been responsible to choose an assertion library by themselves. While this provides freedom it also comes with a burden of extra setup. Thanks to @mgrybyk we now have a custom assertion library that is configured to work with WebdriverIO. In order to avoid the setup burden we should embed this assertion library.

## Types of changes

[//]: # (What types of changes does your code introduce to WebdriverIO?)
[//]: # (_Put an `x` in the boxes that apply_)

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation update

## Checklist

[//]: # (_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code._)

- [x] I have read the [CONTRIBUTING](https://github.com/webdriverio/webdriverio/blob/master/CONTRIBUTING.md) doc
- [x] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)
- [ ] I have added proper type definitions for new commands (if appropriate)

## Further comments

Missing pieces:

- [ ] (requirement) [`expect-webdriverio`](https://github.com/mgrybyk/expect-webdriverio) needs to be moved into the WebdriverIO GitHub org to avoid busfactor effects
- [x] documentation page about general concepts of this library
- [x] API reference in API docs

### Reviewers: @webdriverio/project-committers

mgrybyk commented on examples/wdio/jasmine/jasmine.spec.js:
```diff
@@ -1,7 +1,7 @@
         browser.url('https://webdriver.io')
-        const title = browser.getTitle()
-        expect(title).toBe('WebdriverIO · Next-gen WebDriver test framework for Node.js')
+        expect(browser.getTitle())
```
We can use here expect(browser).toHaveTitle(...)

christian-bromann replied:
@mgrybyk it fails with `TypeError: expect(...).toHaveTitle is not a function`, when running:

```
describe('webdriver.io page', () => {
    it('should have the right title', () => {
        browser.url('https://webdriver.io')
        expect(browser).toHaveTitle('WebdriverIO · Next-gen WebDriver test framework for Node.js')
    })
})
```

mgrybyk replied:
I'll take a look

christian-bromann commented:
This PR is done and can be reviewed. In order to merge this we need to move `expect-webdriverio` to the webdriverio GitHub org, can you take care of it @mgrybyk ?

christian-bromann commented:
Here is the screenshot for the API docs:

<img width="1505" alt="Screenshot 2019-12-25 at 18 33 00" src="https://user-images.githubusercontent.com/731337/71449210-c1c0fb80-2748-11ea-9fe2-34b05f987355.png">

mgrybyk commented:
I'll review everything related to that in next couple of days or so

erwinheitzman commented on docs/Assertion.md:
```diff
@@ -0,0 +1,24 @@
+title: Assertion
+---
+
+> __Note:__ This document is only valid for WebdriverIO > v6.x. If you run an older version of WebdriverIO please update your version or use a 3rd party assertion library like [Chai](https://www.chaijs.com/).
```
We could allow them to install the package separately

erwinheitzman commented on examples/pageobject/specs/checkbox.spec.js:
```diff
@@ -1,18 +1,16 @@
         CheckboxPage.open()
-        assert.equal(CheckboxPage.firstCheckbox.isSelected(), false)
-        assert.equal(CheckboxPage.lastCheckbox.isSelected(), true)
+        expect(CheckboxPage.firstCheckbox).toBeSelected()
```
firstCheckbox should not be selected

erwinheitzman commented on examples/pageobject/specs/checkbox.spec.js:
```diff
@@ -1,18 +1,16 @@
     it('checkbox 1 should be enabled after clicking on it', () => {
         CheckboxPage.open()
-        assert.equal(CheckboxPage.firstCheckbox.isSelected(), false)
+        expect(CheckboxPage.firstCheckbox).toBeSelected()
```
firstCheckbox should not be selected

erwinheitzman commented on examples/pageobject/wdio.conf.js:
```diff
@@ -33,7 +33,7 @@ exports.config = {
     // Level of logging verbosity: trace | debug | info | warn | error | silent
     logLevel: 'trace',
-    outputDir: __dirname,
+    // outputDir: __dirname,
```
Why is this commented?

erwinheitzman commented on packages/wdio-runner/package.json:
```diff
@@ -34,6 +34,7 @@
     "@wdio/logger": "6.0.0-alpha.0",
     "@wdio/utils": "6.0.0-alpha.0",
     "deepmerge": "^4.0.0",
+    "expect-webdriverio": "^0.1.3-beta.7",
```
If we are going to support this I suggest we update the package to no longer include the word `beta`

mgrybyk replied:
updated version to 0.2.0 and transferred repo to webdriverio organization. 

@christian-bromann what are the next steps?

christian-bromann replied:
This PR needs an approval and can then be merged.

christian-bromann commented:
> If we are going to support this I suggest we update the package to no longer include the word `beta`

@mgrybyk did you have a chance to review this and eventually move `expect-webdriverio` into the WebdriverIO GitHub org? This would be the last requirement to get this PR merged.

@erwinheitzman addressed your comments.

mgrybyk commented:
Yes, today will review everything

christian-bromann commented:
@erwinheitzman outstanding issues have been resolved:

- repo has been moved into the org: https://github.com/webdriverio/expect-webdriverio
- a new version was released https://github.com/webdriverio/expect-webdriverio/commit/a01b1dd4981aeb4cbe993f93639762da170baa4a

Anything left to do?
